src/main/java/soc/client/MessageHandler.java

```diff
@@ -0,0 +1,2879 @@
+		pcl.requestedTradeReset(player);
+	}
+
+	/**
+	 * handle the "development card action" message
+	 * @param mes  the message
+	 */
+	protected void handleDEVCARDACTION(final boolean isPractice, final SOCDevCardAction mes)
+	{
+		SOCGame ga = client.games.get(mes.getGame());
+
+		if (ga != null)
+		{
+			final int mesPN = mes.getPlayerNumber();
+			SOCPlayer player = ga.getPlayer(mesPN);
+
+			int ctype = mes.getCardType();
+			if ((! isPractice) && (client.sVersion < SOCDevCardConstants.VERSION_FOR_NEW_TYPES))
+			{
+				if (ctype == SOCDevCardConstants.KNIGHT_FOR_VERS_1_X)
+					ctype = SOCDevCardConstants.KNIGHT;
+				else if (ctype == SOCDevCardConstants.UNKNOWN_FOR_VERS_1_X)
+					ctype = SOCDevCardConstants.UNKNOWN;
+			}
+
+			final int act = mes.getAction();
+			switch (act)
+			{
+			case SOCDevCardAction.DRAW:
+				player.getInventory().addDevCard(1, SOCInventory.NEW, ctype);
+				break;
+
+			case SOCDevCardAction.PLAY:
+				player.getInventory().removeDevCard(SOCInventory.OLD, ctype);
+				// JM temp debug:
+					if (ctype != mes.getCardType())
+						System.out.println("L3947: play dev card type " + ctype + "; srv has " + mes.getCardType());
+					break;
+
+			case SOCDevCardAction.ADD_OLD:
+				player.getInventory().addDevCard(1, SOCInventory.OLD, ctype);
+				break;
+
+			case SOCDevCardAction.ADD_NEW:
+				player.getInventory().addDevCard(1, SOCInventory.NEW, ctype);
+				break;
+			}
+
+			PlayerClientListener pcl = client.getClientListeners().get(mes.getGame());
+			pcl.playerDevCardUpdated(player, (act == SOCDevCardAction.ADD_OLD));
+		}
+	}
+
+	/**
+	 * handle the "set played dev card flag" message
+	 * @param mes  the message
+	 */
+	protected void handleSETPLAYEDDEVCARD(SOCSetPlayedDevCard mes)
+	{
+		SOCGame ga = client.games.get(mes.getGame());
+
+		if (ga != null)
+			SOCDisplaylessPlayerClient.handlePLAYERELEMENT_simple
+			(ga, null, mes.getPlayerNumber(), SOCPlayerElement.SET,
+					SOCPlayerElement.PLAYED_DEV_CARD_FLAG, mes.hasPlayedDevCard() ? 1 : 0, null);
+	}
+
+	/**
+	 * handle the "list of potential settlements" message
+	 * @param mes  the message
+	 * @throws IllegalStateException if the board has
+	 *     {@link SOCBoardLarge#getAddedLayoutPart(String) SOCBoardLarge.getAddedLayoutPart("AL")} != {@code null} but
+	 *     badly formed (node list number 0, or a node list number not followed by a land area number).
+	 *     This Added Layout Part is rarely used, and this would be discovered quickly while testing
+	 *     the board layout that contained it.
+	 */
+	protected void handlePOTENTIALSETTLEMENTS(SOCPotentialSettlements mes)
+			throws IllegalStateException
+	{
+		System.err.println("L3292 potentialsettles at " + System.currentTimeMillis());
+		SOCDisplaylessPlayerClient.handlePOTENTIALSETTLEMENTS(mes, client.games);
+
+		PlayerClientListener pcl = client.getClientListeners().get(mes.getGame());
+		if (pcl != null)
+			pcl.boardPotentialsU
```

Comment:
This was {@link SwingMainDisplay.GameOptionsTimeoutTask} and probably should still be that; probably a search-and-replace typo

Reply:
^ although github's preview shows line 2198, i made this comment at 2400, which is:
`{@link Swingclient.getMainDisplay().GameOptionsTimeoutTask} can eventually call this`